### PR TITLE
[WIP] Adding Support for Foreign Buffers

### PR DESCRIPTION
**Pull request by mt-jones**
This PR closes https://github.com/dmlc/xgboost/pull/3997 in favor of a design which removes cuDF as a dependency for handling foreign devices buffers.

### CODE REVIEW

File: tests/python-gpu/test_gpu_gdf.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
+import numpy as np
+import pandas as pd
+try:
+    import cudf.dataframe as gdf
```
Comment (mt-jones) [inline tests/python-gpu/test_gpu_gdf.py]: cuDF references in Python need to be removed, and reworked to accept two cases:
1. a list of `__cuda_array_interface__` objects ... equivalent to a set of columns
2. a single `__cuda_array_interface__` object ... equivalent to accepting a matrix of values

File: src/data/simple_csr_source.cu [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+      CountValid <<<blocks, threads>>> (cols[i]->valid, n_rows, n_cols, csr->offsets);
+      dh::safe_cuda(cudaGetLastError());
+      dh::safe_cuda(cudaDeviceSynchronize());
+    }
+
+    thrust::device_ptr<size_t> offsets(csr->offsets);
+    int64_t n_valid = thrust::reduce(offsets, offsets + n_rows, 0ull, thrust::plus<size_t>());
+    thrust::exclusive_scan(offsets, offsets + n_rows + 1, offsets);
+
+    csr->n_nonzero = n_valid;
```
Comment (mt-jones) [inline src/data/simple_csr_source.cu]: Calling out the fact that the intermediate `ForeignCSR` uses the number of _valid_ entries for the number of nonzero elements.

**Comment by trivialfis** [comment]
Hi @mt-jones ,

I saw that you use `data[col]._column._pointer` as the input pointer:
``` python
        if isinstance(data, CUDF):
            col_ptrs = [ctypes.c_void_p(data[col]._column._pointer) for col in data.columns]
```

In my testing script:
``` python
import numpy as np
import pandas as pd
import cudf
# import xgboost as xgb
import ctypes

kRows = 10
kCols = 2

na = np.random.randn(kRows, kCols)
# na[3, 3] = np.NAN
pa = pd.DataFrame(na)

names = []

for i in range(0, kCols):
    names.append(str(i))
pa.columns = names

cd: cudf.DataFrame = cudf.from_pandas(pa)


def print_ptrs(it):
    print('iter: ', it)
    for col in cd.columns:
        print('ptr:', ctypes.c_void_p(cd[col]._column._pointer))
    print()


for i in range(10):
    print_ptrs(i)
```

It generates the output sequence:
```
iter:  1
ptr: c_void_p(2566349056)
ptr: c_void_p(2566235584)

iter:  2
ptr: c_void_p(2566353552)
ptr: c_void_p(2572078624)

iter:  3
ptr: c_void_p(2565799680)
ptr: c_void_p(2568010480)

iter:  4
ptr: c_void_p(2576891344)
ptr: c_void_p(2566280544)

iter:  5
ptr: c_void_p(2566295040)
ptr: c_void_p(2565610112)

iter:  6
ptr: c_void_p(2566397088)
ptr: c_void_p(2573274528)

iter:  7
ptr: c_void_p(2568859520)
ptr: c_void_p(2577131840)

iter:  8
ptr: c_void_p(2573920752)
ptr: c_void_p(2566281040)

iter:  9
ptr: c_void_p(2572178928)
ptr: c_void_p(2568486224)
```

So apparently this pointer is changing at every access?  Do you have any suggestion for accessing underlying data pointer?

**Comment by trivialfis** [comment]
Suppressed by #4745.

**Pull request closed without merging**